### ISSUE TITLE
UIManager.getSytemLookAndFeelClassName() returns wrong value on Fedora 32

### ISSUE THREAD

**Issue by tresf** [adoptium/adoptium-support#94]
**Description**

* As identified in https://github.com/AdoptOpenJDK/openjdk-support/issues/4#issuecomment-620661519, newer Fedora versions don't know to use the Gnome/Gtk theme.  This bug may also affect other Gtk-based desktops which do not provide this value.

* Upstream Gnome commit which removed `GNOME_DESKTOP_SESSION_ID`: https://gitlab.gnome.org/GNOME/gnome-session/commit/00e0e6226371d53f651cc881e74c0543192c94a8#5b3005b925ed5c2612a9604ad3c756b1f9472165

* Question to Gnome's desktop-devel-list: https://mail.gnome.org/archives/desktop-devel-list/2020-April/msg00012.html

**Platform and architecture**:
* Fedora 32 (Workstation) x86_64

**Java version**:
```
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.7+10, mixed mode)
```

**Actual vs. Expected Behavior**:
```java
System.out.println(UIManager.getSystemLookAndFeelClassName());
```

   * On Fedora 32, it returns: (**ACTUAL**)
```
javax.swing.plaf.metal.MetalLookAndFeel
```

   * On Fedora 27, it returns: (**EXPECTED**)
```
com.sun.java.swing.plaf.gtk.GTKLookAndFeel
```

**Details**

Gnome provides an environment variable `GNOME_DESKTOP_SESSION_ID=this-is-deprecated`, however newer Fedora versions do not provide this environment variable.  More details on which process sets this variable is available here: https://stackoverflow.com/q/40603418/3196753

OpenJDK looks for this variable in [`UNIXToolkit:L100`](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/634c676e6c145c936b460debaa95d0b6cfa61ae2/src/java.desktop/unix/classes/sun/awt/UNIXToolkit.java#L100)

**Comment by aahlenst** [comment]
I didn't find a JBS entry that matches your description.

@jerboaa Do you need more info?

**Comment by tresf** [comment]
A bit more information...  This isn't a Wayland issue, but rather the removal of the value from the Gnome desktop.  Evidence is that this value exists on Fedora 27 using Wayland and X11 but is missing on Fedora 32 using X11 and Wayland.  Below is a detailed matrix.

One fallback is to trust `$XDG_CURRENT_DESKTOP`, however this value seems to be less reliable for Gnome-based desktop environments.


| OS | `$XDG_CURRENT_DESKTOP` | `$GNOME_DESKTOP_SESSION_ID` | OK? |
|-----|------------------------------|-------------------------------------|----------|
| Solaris 11 | `GNOME` | `this-is-deprecated`| ✅ |
| Solaris 10 | <kbd>blank</kbd> | `Default`| ✅|
| Ubuntu Server 18.04 (Headless) |<kbd>blank</kbd> |<kbd>blank</kbd> | ✅|
| Ubuntu Desktop 18.04 (X11) | `ubuntu:GNOME` | `this-is-deprecated`| ✅|
| Ubuntu Desktop 18.04 (Wayland) | `ubuntu:GNOME` | `this-is-deprecated`| ✅|
| Ubuntu Desktop 16.04 (X11) | `Unity` | `this-is-deprecated`| ✅|
| Ubuntu Desktop 14.04 (X11) | `Unity` | `this-is-deprecated`| ✅|
| Fedora 32 (X11) | `GNOME` |<kbd>blank</kbd>| ⛔ |
| Fedora 32 (Wayland) | `GNOME` |<kbd>blank</kbd>|  ⛔|
| Fedora 27 (X11) | `GNOME` | `this-is-deprecated` | ✅|
| Fedora 27 (Wayland) | `GNOME` | `this-is-deprecated` | ✅|
| Debian 9.5 | `GNOME` | `this-is-deprecated`| ✅|
| Manjaro | `XFCE` |<kbd>blank</kbd>| ⛔ | ✅|
| FreeBSD 11.2 | `GNOME` | `this-is-deprecated` | ✅|
| ElementaryOS 5.0 | `Pantheon` | `this-is-deprecated` |
| Haiku R1B1 |<kbd>blank</kbd> |<kbd>blank</kbd> | ✅|

**Comment by tresf** [comment]
I've posted a question to the Gnome Desktop mailing list to see what their recommendation is.  Once it hits the archives, I can hard-link it here.

**Comment by tresf** [comment]
Gnome desktop-devel-list question permalink: https://mail.gnome.org/archives/desktop-devel-list/2020-April/msg00012.html

**Comment by tresf** [comment]
Upstream gnome commit which removed the `GNOME_DESKTOP_SESSION_ID` value: https://gitlab.gnome.org/GNOME/gnome-session/commit/00e0e6226371d53f651cc881e74c0543192c94a8#5b3005b925ed5c2612a9604ad3c756b1f9472165

**Comment by tresf** [comment]
I've found some Debian code using the following code:

```c
os.getenv("XDG_CURRENT_DESKTOP", "").endswith("GNOME") 
```

... which per above table will break on many 3rd party Gtk-based desktops.  I'm hoping Gnome offers some insight into how to do this.  At the time of the above commit hitting, Debian had 225 instances of `GNOME_DESKTOP_SESSION_ID` being used for various desktop environment purposes.  It seems a suitable replacement is warranted and there's hundreds of use-cases to support that.

**Comment by tresf** [comment]
Here's a Gnome project using similar logic: https://gitlab.gnome.org/GNOME/glib-networking/commit/89c45b82b85daef0f1e679dbd15a9c7c893deb17#d53b9ec213c7c4dd230c7ad15ccf62b542a79279_316_316

**Comment by tresf** [comment]
In short, the Gnome developers recommend just loading the theme -- rather blindly -- on all Linuxes.  I'm not sure how the OpenJDK team will react to this proposal, but long of it's all spelled out in the email back-and-forth https://github.com/AdoptOpenJDK/openjdk-support/issues/94#issuecomment-621404508.

I'll quote the relevant parts:

> > This would mean each Gtk-based DE will require a patch to OpenJDK,
> which is unsustainable.

> I agree that this is inconvenient but in the absence of a better
option, you either make a list, or come up with some heuristic.

> I'm not sure I understand the reluctance to provide such information.

> > I do not think anyone is opposed to that but I doubt anyone will
consider it a priority. Especially since the intersection of
toolkits supported by AWT and toolkits prefered can be emulated using
much simpler algorithm. (Gtk|KDE|Motif chart provided).

**Comment by jtojnar** [comment]
The main point is, even if there was API returning the preferred toolkit for the desktop environment, you could rewrite

```java
if (Environment.getPreferredToolkit() == GTK) {
    return "com.sun.java.swing.plaf.gtk.GTKLookAndFeel";
} else if (Environment.getPreferredToolkit() == QT) {
    // Qt not supported by AWT so let's check for GTK as the next best thing
    if (Environment.isGtkAvailable()) {
        return "com.sun.java.swing.plaf.gtk.GTKLookAndFeel";
    }
    return getCrossPlatformLookAndFeelClassName();
} else {
    // GTK is still more user friendly then Metal and the user already has it available
    if (Environment.isGtkAvailable()) {
        return "com.sun.java.swing.plaf.gtk.GTKLookAndFeel";
    }
    return getCrossPlatformLookAndFeelClassName();
}
```

into

```java
if (Environment.isGtkAvailable()) {
    return "com.sun.java.swing.plaf.gtk.GTKLookAndFeel";
}
return getCrossPlatformLookAndFeelClassName();
```

----

Also need to point out that I am not a member or representative of GNOME project, I only package GNOME and Freedesktop projects for NixOS.

**Comment by tresf** [comment]
I've filed this bug with our JDK support provider.  If they file upstream before an Adopt dev gets around to it, I'll link.

**Comment by tresf** [comment]
The upstream bug report was filed but closed. https://bugs.openjdk.java.net/browse/JDK-8247753

Quoting:

> We only support Ubuntu and Oracle Linux. We don't support other flavours of Linux, so this will be closed unless this is reproducible on these two platforms.
> Please re-open if required.

This culture terrifies me. 🤕

**Comment by karianna** [comment]
I got it reopened upstream

**Comment by tresf** [comment]
Update: It appears OpenJDK will be accepting the `XDG_CURRENT_DESKTOP` patch.  This still has scalability issues, but addresses the problem on the vast majority of distributions.

I'm uncertain if this is targeted at JDK11 (the mail says JDK16) or perhaps if it will be backported. Here's the mailing list conversation: https://mail.openjdk.java.net/pipermail/swing-dev/2020-August/010608.html

**Comment by tresf** [comment]
The patch has landed on JDK16, just awaiting the backport.

https://bugs.openjdk.java.net/browse/JDK-8247753

**Comment by aahlenst** [comment]
@tresf I haven't seen any activity regarding a backport to 11. Do you know more? Or has someone else to step in!

**Comment by jerboaa** [comment]
Waiting on review upstream: https://mail.openjdk.java.net/pipermail/jdk-updates-dev/2021-February/005078.html

**Comment by karianna** [comment]
Fixed in 11.0.12
